### PR TITLE
First and third party cookie blocking

### PR DESCRIPTION
**Pull request by sammacbeth**
**Reviewer:**

## Description:

This PR merges the changes from #540  and #545, and syncs them with the latest develop branch. 

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/chrome-events.es6.js [hunk 1]
```diff
@@ -12,6 +13,8 @@ const browser = utils.getBrowserName()
 
 const sha1 = require('../shared-utils/sha1')
 
+const IS_BETA = chrome.runtime.id !== 'bkdgflcldnnnapblkhphbgpggdiikppg' && chrome.runtime.id !== 'jid1-ZAdIEUB7XOzOJw@jetpack'
```
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:16]: This will likely break on edge - we should add the edge extension ID here.

File: shared/js/background/utils.es6.js [hunk 2]
```diff
@@ -150,14 +150,16 @@ function getAsyncBlockingSupport () {
...
 */
 function isBroken (url) {
-    if (!tdsStorage?.brokenSiteList) return
+    if (!tdsStorage || !tdsStorage.brokenSiteList) return
     return isBrokenList(url, tdsStorage.brokenSiteList)
 }
 
 function isCanvasBroken (url) {
-    if (!tdsStorage?.fingerprinting?.canvas?.sites) return
+    if (!tdsStorage || !tdsStorage.fingerprinting || !tdsStorage.fingerprinting.canvas) return
```
Comment (jonathanKingston) [inline shared/js/background/utils.es6.js:158]: It seems like these changes here came in with 19bd1fe7b67da32c1df051100ecabc98bddedd32 can we remove them? Or were they intended?

File: shared/js/content-scripts/block-cookie.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
+/* global Cookie */
+// Set up 3rd party cookie blocker
+(function cookieBlocker () {
+    // don't inject into non-HTML documents (such as XML documents)
+    // but do inject into XHTML documents
+    if (document instanceof HTMLDocument === false && (
+        document instanceof XMLDocument === false ||
+        document.createElement('div') instanceof HTMLDivElement === false
```
Comment (jonathanKingston) [inline shared/js/content-scripts/block-cookie.js:8]: The page could game this to return a different prototype.

File: shared/js/background/utils.es6.js (hunk 2 above)
Comment (sammacbeth) [reply shared/js/background/utils.es6.js:158]: @SlayterDev did you have some issue with optional chaining that prompted you to change this?

@jonathanKingston I assume this gets transpiled as we target a firefox min version of 57, which is before support for this was added https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: shared/js/content-scripts/block-cookie.js (hunk 3 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:8]: This check is run in content-script scope, where that cannot happen, right?

File: shared/js/background/utils.es6.js (hunk 2 above)
Comment (jonathanKingston) [reply shared/js/background/utils.es6.js:158]: Yeah this this super easy to read statement:
```
return (tdsStorage === null || tdsStorage === void 0 ? void 0 : (_tdsStorage$fingerpri2 = tdsStorage.fingerprinting) === null || _tdsStorage$fingerpri2 === void 0 ? void 0 : (_tdsStorage$fingerpri3 = _tdsStorage$fingerpri2.canvas) === null || _tdsStorage$fingerpri3 === void 0 ? void 0 : _tdsStorage$fingerpri3.scripts) || [];
```

File: shared/js/content-scripts/block-cookie.js (hunk 3 above)
Comment (jonathanKingston) [reply shared/js/content-scripts/block-cookie.js:8]: Yeah sorry I thought this was injected :)

File: shared/js/background/utils.es6.js (hunk 2 above)
Comment (SlayterDev) [reply shared/js/background/utils.es6.js:158]: Yes I made this change. I was having issues with linting with the optional chain and @kdzwinel suggested we refactor it as optional chaining isn't well supported in the linter yet.

Comment (jonathanKingston) [reply shared/js/background/utils.es6.js:158]: The linter should support this now after updating the the latest "standard"

Comment (SlayterDev) [reply shared/js/background/utils.es6.js:158]: Could've been. I have no issue switching it back as the optionals were much cleaner. I'll give it a shot.

File: shared/js/background/chrome-events.es6.js [hunk 4]
```diff
@@ -214,6 +279,44 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
...
+        }
+        if (chrome.runtime.lastError) { // Prevent thrown errors when the frame disappears
+            return true
+        }
+        if (blockingExperimentActive()) {
+            const tab = tabManager.get({ tabId: sender.tab.id })
+            // determine the register domain of the sending tab
+            const tabUrl = tab ? tab.url : sender.tab.url
+            const parsed = tldts.parse(tabUrl)
+            action.tabRegisteredDomain = parsed.isIp ? parsed.hostname : parsed.domain || parsed.hostname
```
Comment (kdzwinel) [inline shared/js/background/chrome-events.es6.js:299]: in what cases url is not an IP but doesn't have domain? (why we have to do `|| parsed.hostname` part?)

File: shared/js/background/chrome-events.es6.js [hunk 5]
```diff
@@ -214,6 +279,44 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
...
+
+            if (req.documentUrl && trackerutils.isTracker(req.documentUrl) && sender.frameId !== 0) {
+                action.isTrackerFrame = true
+            }
+
+            if (tab && tab.site.whitelisted) {
```
Comment (kdzwinel) [inline shared/js/background/chrome-events.es6.js:305]: could we do this check first? i.e. before parsing tabUrl and checking if frame is tracking?

File: shared/js/background/chrome-events.es6.js [hunk 6]
```diff
@@ -214,6 +279,44 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
...
+                res(action)
+                return true
+            }
+
+            action.isThirdParty = !utils.isFirstParty(sender.url, sender.tab.url)
+            action.shouldBlock = !cookieConfig.isExcluded(sender.url)
+            res(action)
+        } else {
+            res(action)
+            return true
```
Comment (kdzwinel) [inline shared/js/background/chrome-events.es6.js:315]: nit: we can skip the return here

File: shared/js/background/utils.es6.js [hunk 7]
```diff
@@ -201,20 +201,35 @@ function isSafeListed (url) {
...
 
+/**
+ * Tests whether the two URL's belong to the same
+ * first party set.
+ */
+function isFirstParty (url1, url2) {
+    const first = tldts.parse(url1)
+    const second = tldts.parse(url2)
+
+    const firstDomain = first.isIp ? first.hostname : first.domain
```
Comment (kdzwinel) [inline shared/js/background/utils.es6.js:212]: we are not doing the `|| hostname` thing here 👀 should we?

File: shared/js/content-scripts/block-cookie.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    )) {
+        return
+    }
+
+    function clearInjectedCookiesAndBlock () {
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:13]: this function name no longer makes sense (we don't clear cookies) - can we update?

File: shared/js/content-scripts/block-cookie.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                        debug && console.log('[ddg-cookie-policy] ignored (non-tracker)', value, [...scriptOrigins])
+                        return
+                    }
+                    // extract cookie expiry from cookie string
+                    const cookie = new Cookie(value)
+                    // apply cookie policy
+                    if (cookie.getExpiry() > policy.threshold) {
+                        cookie.maxAge = policy.maxAge
+                        debug && console.log('[ddg-cookie-policy] update', cookie.toString(), scriptOrigins)
+                        cookieSetter.apply(document, [cookie.toString()])
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:83]: there is a slight chance for a race condition where cookie was removed in the meantime and we are recreating it. Maybe it's worth checking if it's still in `document.cookie` before we write?

**Review by kdzwinel — COMMENTED** [review]
Awesome work ❤️ I haven't found anything critical, but I have left bunch of questions.

File: shared/data/constants.js [hunk 10]
```diff
@@ -123,6 +123,14 @@ module.exports = {
             "source": "external"
         }
     ],
+    "CookieLists": [
+        {
+            "name": "cookieExcludeList",
+            "url": "https://staticcdn.duckduckgo.com/useragents/cookie_configuration.json",
```
[the remark this answers is not in the document]
Comment (sammacbeth) [reply shared/data/constants.js:129]: There are config options for 1p cookies in there, though I guess not deployed yet: https://dub.duckduckgo.com/duckduckgo/UserAgent-Deployments/commit/0c6df3f7cb3dddb6634a4a8ada60d9842cfecb92

File: shared/js/background/chrome-events.es6.js [hunk 11]
```diff
@@ -12,6 +13,13 @@ const browser = utils.getBrowserName()
 
 const sha1 = require('../shared-utils/sha1')
 
+const RELEASE_EXTENSION_IDS = [
+    'caoacbimdbbljakfhgikoodekdnlcgpk', // edge store
+    'bkdgflcldnnnapblkhphbgpggdiikppg', // chrome store
+    'jid1-ZAdIEUB7XOzOJw@jetpack' // firefox
+]
+const IS_BETA = RELEASE_EXTENSION_IDS.indexOf(chrome.runtime.id) === -1
```
[the remark this answers is not in the document]
Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js:21]: Looks like this is a relic of old extensions build for addons SDK: https://wiki.mozilla.org/Jetpack

File: shared/js/background/chrome-events.es6.js [hunk 12]
```diff
@@ -60,6 +85,32 @@ chrome.webRequest.onBeforeRequest.addListener(
...
 
+/**
+ * Listen for script loads from trackers and relay the hosts to the frame's content-script.
+ * This will be used for tracker-dependant decicions in that context.
+ */
+chrome.webRequest.onBeforeRequest.addListener(
+    (details) => {
+        if (trackerutils.isTracker(details.url)) {
+            console.log('tracker script', details.url)
+            chrome.tabs.sendMessage(details.tabId, {
```
[the remark this answers is not in the document]
Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js:96]: We could move the code into `handleRequest` in `redirect.es6.js`. However, this may get messy as there are several branches in that code, so to send the message for only the case when we do not block or redirect but it is a tracker may made the code more difficult to read.

File: shared/js/background/chrome-events.es6.js (hunk 4 above)
Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js:299]: `localhost` and any other hostname without a `.`.

File: shared/js/background/chrome-events.es6.js (hunk 5 above)
Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js:305]: Makes sense - I think this ordering is a consequence of merging the separate changes.

File: shared/js/background/utils.es6.js (hunk 7 above)
Comment (sammacbeth) [reply shared/js/background/utils.es6.js:212]: Looks like it - I added tests and there is an edge case that fails due to this: https://github.com/duckduckgo/duckduckgo-privacy-extension/runs/2164976053?check_suite_focus=true#step:5:296

Should we also using private PSL entries here too to align origin boundaries with the browser?

File: shared/js/content-scripts/block-cookie.js (hunk 9 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:83]: Good catch. I'll add a check.

File: shared/js/background/utils.es6.js (hunk 7 above)
Comment (sammacbeth) [reply shared/js/background/utils.es6.js]: Fixed by https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/570/commits/b1763ff0f218447493bfa68d62164b55a2abef51

File: shared/js/content-scripts/block-cookie.js (hunk 9 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js]: Fixed by https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/570/commits/862fbf6a595bc66f17d272b9c17ab4c2288b6227

File: shared/data/experiments-out.js [hunk 13]
```diff
@@ -15,5 +15,20 @@ module.exports = {
                 }
             }
         }
+    },
+    o: {
```
[the remark this answers is not in the document]
Comment (sammacbeth) [reply shared/data/experiments-out.js:19]: @SlayterDev could you have a look at this? I'm not sure how the test names are set up.

File: shared/js/background/chrome-events.es6.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js]: @kdzwinel so the change needed looks to be ok actually: https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/570/commits/6446343dff44899f54365cd2818af84c78711a6f

File: shared/data/experiments-out.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (SlayterDev) [reply shared/data/experiments-out.js:19]: As I understand it this needs to be 'o' for the experiment to work. I have updated the description though.

File: shared/data/constants.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (SlayterDev) [reply shared/data/constants.js:129]: Appears deployments were broken on this repo. I've pushed a fix so this should get updated soon.
